File: Exercism Codes/PHP/BookStore.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

/**
 * Note: we expect the total in cents (1$ = 100 cents).
 */
function total(array $items): int
{
    $cnt = getFrequency($items);
    $freq = !empty(array_values($cnt)) ? max(array_values($cnt)) : 0;
    $combinations = findCombinations($cnt, $freq, array_keys($cnt), 0, [array_fill(0, $freq, 0)]);
    $discounts = [1 => 0.0, 2 => 0.4, 3 => 0.8, 4 => 1.6, 5 => 2.0];

    return min(array_map(function($lst) use ($discounts) {
        return array_sum(array_map(function($b) use ($discounts) {
            return intval(100 * $b * (8.0 - $discounts[$b]));
        }, $lst));
    }, $combinations));
}

function getFrequency(array $books): array
{
    $frequency = [];
    foreach ($books as $x) {
        if (array_key_exists($x, $frequency)) {
            $frequency[$x] += 1;
        }
        else {
            $frequency[$x] = 1;
        }
    }
    return $frequency;
}

function findCombinations(array $cnt, int $freq, array $nums, int $num, array $arr) 
{
    if ($num == count($nums)) {
        return $arr;
    }
    else {
        $arr2 = $arr;
        $arr = [];

        $positionsList = getCombinations(range(0, $freq - 1), $cnt[$nums[$num]]);
        foreach ($arr2 as $x) {
            foreach ($positionsList as $positions) {
                $newArr = $x;
                foreach ($positions as $j) {
                    $newArr[$j] += 1;
                }
                $arr[] = $newArr;
            }
        }
        return findCombinations($cnt, $freq, $nums, $num + 1, $arr);
    }
}

function getCombinations(array $arr, int $r): array
{
    $result = [];

    $backtrack = function(int $start, array $current) use (&$result, $arr, $r, &$backtrack) {
        if (count($current) == $r) {
            $result[] = $current;
            return;
        }
        for ($i = $start; $i < count($arr); $i++) {
            $current[] = $arr[$i];
            $backtrack($i + 1, $current);
            array_pop($current);
        }
    };
    $backtrack(0, []);
    return $result;
}

File: Exercism Codes/C#/BookStore.cs
public static class BookStore
{
    public static decimal Total(IEnumerable<int> books)
    {
        IDictionary<int, int> cnt = GetFrequency(books);
        int freq = cnt.Values.DefaultIfEmpty(0).Max();
        List<List<int>> combinations = FindCombinations(
            cnt, freq, cnt.Keys.ToList(), 0, new List<List<int>> { Enumerable.Repeat(0, freq).ToList() }
            );
        IDictionary<int, decimal> discounts = new Dictionary<int, decimal>
        {
            { 1, 0.0m },
            { 2, 0.4m },
            { 3, 0.8m },
            { 4, 1.6m },
            { 5, 2.0m }
        };
        return combinations.Select(lst => lst.Select(books => books * (8.0m - discounts[books])).Sum()).Min();
    }

    private static IDictionary<int, int> GetFrequency(IEnumerable<int> books)
    {
        Dictionary<int, int> frequency = new Dictionary<int, int>();

        foreach (int x in books)
        {
            if (frequency.ContainsKey(x))
            {
                frequency[x]++;
            }
            else
            {
                frequency[x] = 1;
            }
        }

        return frequency;
    }

    private static List<List<int>> FindCombinations(IDictionary<int, int> cnt, int freq, List<int> nums, int num, List<List<int>> arr)
    {
        if (num == nums.Count)
        {
            return arr;
        }
        else
        {
            List<List<int>> arr2 = new List<List<int>>(arr);
            arr = new List<List<int>>();

            foreach (List<int> x in arr2)
            {
                foreach (List<int> positions in GetCombinations(Enumerable.Range(0, freq).ToList(), cnt[nums[num]]))
                {
                    List<int> newCombination = new List<int>(x);
                    foreach (int pos in positions)
                    {
                        newCombination[pos] += 1;
                    }
                    arr.Add(newCombination);
                }
            }

            return FindCombinations(cnt, freq, nums, num + 1, arr);
        }
    }

    private static List<List<int>> GetCombinations(List<int> arr, int r)
    {
        List<List<int>> result = new List<List<int>>();

        void Backtrack(int start, List<int> current)
        {
            if (current.Count == r)
            {
                result.Add(new List<int>(current));
                return;
            }
            for (int i = start; i < arr.Count; i++)
            {
                current.Add(arr[i]);
                Backtrack(i + 1, current);
                current.RemoveAt(current.Count - 1);
            }
        }

        Backtrack(0, new List<int>());
        return result;
    }
}